### DISCUSSION TITLE
Fix buggy taskbar stacking (z-index) behavior

### DISCUSSION THREAD

**Discussion by stamminator** [ramensoftware/windhawk-mods#1205] [General]
When the Windows 11 start menu is open, the taskbar gets stacked on top of all other UI elements, which I'll call "bring taskbar to front" mode or BTTF for brevity. When in this mode, several defects take place:

1. Right clicking on notification tray icons results in popup menus that get displayed _behind_ the taskbar, making some menu options unclickable. This happens for all apps I've tested.

![Animation1](https://github.com/user-attachments/assets/5d551e76-4938-4134-a14c-04ddf2c70cf5)

2. Normally, any open tray icon context menus get closed when the user clicks anywhere other than that menu, which is standard popup behavior. But any tray icon context menus that get opened while BTTF mode is active end up staying open and can only be closed by triggering its close event via the app itself, such as by selecting one of the menu options or leaving BTTF mode and right-clicking again on the tray icon. This happens for all apps I've tested except for Monitorian, whose context menu manages to still respond to system-sent close events even if it was opened in BTTF mode.

![Animation2](https://github.com/user-attachments/assets/1f2a4695-5d83-4137-b56d-ab10e15dc41b)

3. You can get duplicate instances of an app's popup menu that don't disappear until you terminate that app's process. I only got this behavior with the MagicPods app, so it's likely an issue with the app's code not treating the context menu as a singleton, but still worth showing here.

![Animation3](https://github.com/user-attachments/assets/957c4826-ac39-4b85-b3b9-4e3701802ff8)

Is any of this behavior capable of being fixed via the `windows-11-taskbar-styler` mod? If so, here are some considerations:

- If the implementation were to be "keep the taskbar's ZIndex the same even if the start menu is open", although it would solve the tray icon context menu issue, wouldn't that also prevent the taskbar from appearing when the Win key is pressed while an app is in full screen mode? This would be undesirable.
- Icons in the overflow menu don't experience this issue because right-clicking them automatically closes the start menu and exits "bring taskbar to front" mode. One solution could be to do the same for all tray icons. If so, this may fit better as a toggle setting in the taskbar-`tray-system-icon-tweaks` mod.

**Comment by m417z** [comment]
> Is any of this behavior capable of being fixed via the `windows-11-taskbar-styler` mod?

I don't think so.

> Icons in the overflow menu don't experience this issue because right-clicking them automatically closes the start menu and exits "bring taskbar to front" mode. One solution could be to do the same for all tray icons.

Interesting, I wonder whether it's accidental, other popups (jump lists, taskbar menus, action center, notification center, etc.) don't close the start menu. Closing it when interacting with tray icons would probably be the easiest solution.

**Reply by stamminator** [reply]
> Closing it when interacting with tray icons would probably be the easiest solution.

I agree, although it would be frustrating behavior in some scenarios. Imagine you're playing a full-screen game and want to change your Disord online status. You press the Win key to show the taskbar, right click on the Discord tray icon, and rather than the context menu appearing, the taskbar vanishes.

For this reason, I think the ideal fix would have to consist of the following:

- Interacting with tray icons (including ones in the overflow menu) should not close the start menu
- BTTF mode should still exist, but tray icon context menus should always have a ZIndex of "taskbar's ZIndex + 1".
- BTTF mode should not cause tray icon context menus to be "orphaned". Their auto-closing behavior should be the same whether in or out of BTTF mode.

It sounds like you're saying Windhawk mods aren't capable of achieving this, so the ball is in Microsoft's court. We can expect a fix in Windows 12, build 31H2 🥲

**Reply by m417z** [reply]
> It sounds like you're saying Windhawk mods aren't capable of achieving this

Well, not effortlessly. It will take some research to understand what's possible and the required amount of effort.

**Comment by crossterra-agent** [comment]
- **Status**: DONE
- **Role**: developer
- **Link**: https://github.com/CrossTerra/monorepo/pull/3021
- **Summary**: All steps complete (S01–S05); 20 tests pass, pre-existing action test failures are unrelated to this PR and present on main.
